awx/main/tasks/facts.py

```diff
@@ -0,0 +1,126 @@
+import codecs
+import datetime
+import os
+import json
+import logging
+
+# Django
+from django.conf import settings
+from django.db.models.query import QuerySet
+from django.utils.encoding import smart_str
+from django.utils.timezone import now
+
+# AWX
+from awx.main.utils.common import log_excess_runtime
+
+
+logger = logging.getLogger('awx.main.tasks.facts')
+system_tracking_logger = logging.getLogger('awx.analytics.system_tracking')
+
+
+def _get_inventory_hosts(inventory, slice_number, slice_count, only=('name', 'ansible_facts', 'ansible_facts_modified', 'modified', 'inventory_id'), **filters):
+    """Return value is an iterable for the relevant hosts for this job"""
+    if not inventory:
+        return []
+    host_queryset = inventory.hosts.only(*only)
+    if filters:
+        host_queryset = host_queryset.filter(**filters)
+    host_queryset = inventory.get_sliced_hosts(host_queryset, slice_number, slice_count)
+    if isinstance(host_queryset, QuerySet):
+        return host_queryset.iterator()
+    return host_queryset
+
+
+@log_excess_runtime(logger, debug_cutoff=0.01, msg='Inventory {inventory_id} host facts prepared for {written_ct} hosts, took {delta:.3f} s', add_log_data=True)
+def start_fact_cache(inventory, destination, log_data, timeout=None, slice_number=0, slice_count=1):
+    log_data['inventory_id'] = inventory.id
+    log_data['written_ct'] = 0
+    try:
+        os.makedirs(destination, mode=0o700)
```

Comment:
nit: Can we use `exist_ok=True` and skip the try/except? Was introduced in 3.2

Reply:
The reason the except is needed here but not before is because we loop over `input_inventories`, so the dir will exist after the first one. I've thought that we could refactor this away instead.

Reply:
I think I agree that using `exist_ok=True` is a bit cleaner here. It effectively does the same thing, just avoids throwing the exception at all, so the try/except/pass lines become unnecessary.